presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java

```diff
@@ -59,46 +60,36 @@ public PartitioningExchanger(
     }
 
     @Override
-    public void accept(Page page)
+    public synchronized void accept(Page page)
     {
-        // extract the partitioning channel before entering the critical section
-        partitionPage(page, extractPartitioningChannels(page));
-    }
+        // reset the assignment lists
+        for (IntList partitionAssignment : partitionAssignments) {
+            partitionAssignment.clear();
+        }
 
-    private synchronized void partitionPage(Page page, Page partitioningChannelsPage)
-    {
-        // assign each row to a partition. The assignments lists are all expected to cleared by the previous iterations
+        // assign each row to a partition
+        Page partitioningChannelsPage = extractPartitioningChannels(page);
         for (int position = 0; position < partitioningChannelsPage.getPositionCount(); position++) {
             int partition = partitionFunction.getPartition(partitioningChannelsPage, position);
             partitionAssignments[partition].add(position);
         }
 
         // build a page for each partition
-        for (int partition = 0; partition < partitionAssignments.length; partition++) {
+        for (int partition = 0; partition < buffers.size(); partition++) {
             IntArrayList positions = partitionAssignments[partition];
-            int partitionSize = positions.size();
-            if (partitionSize == 0) {
-                continue;
-            }
-            Page pageSplit;
-            if (partitionSize == page.getPositionCount()) {
-                pageSplit = page; // entire page will be sent to this partition, no copies necessary
```

Comment:
Is it possible that the issue is that these pages were retaining more memory as a result of not having all of their positions explicitly copied out? If this partitioning exchanger was used to do a local exchange after a remote exchange (fairly common), then the issue with [VariableWidthBlocks retaining their entire SerializedPage slice](https://github.com/prestodb/presto/pull/16248) could indeed yield a much higher retained size than before.

Reply:
Sorry about the delayed reply. @aweisberg do you have more context on this since you mentioned this PR in our discussion about the local memory issues?

Reply:
I think I talked about this with James in Slack, but we didn't pick the correct way to follow up and get this performance improvement without also sometimes regressing on memory.

Reply:
Since I had also made the regressing change to Trino, I followed up there with a fix and a longer description of the problem / alternatives. See: https://github.com/trinodb/trino/pull/9327 (with a follow up PR in https://github.com/trinodb/trino/pull/9379)

As for what to do in PrestoDB, there some unknowns about reintroducing this change with a similar fix on the Trino side, as I do not have a comparable testing environment to ensure that calling `Page#compact()` explicitly will perform comparably to `Page#copyPositions` in terms of throughput or reported memory usage for all workloads. Trino has made a variety of other localized changes to related classes (like adding explicit calls to `Page#compact()` in other places, revised operator and memory tracking implementations, etc). On the Athena side, we chose to take the performance hit of eagerly copying `VariableWidthBlock` instances out of their input slice during deserialization which means we didn't have the same issue, but that's an expensive hit to take to throughput.

So as I see it, we can either:
1. Take a harsh performance hit to `VariableWidthBlock` deserialization throughput, but address this basic memory tracking bug more thoroughly (ie: right now the memory tracking bug isn't typically reported because there is _usually_ a partitioning local exchange after a remote exchange that forces the copy, but not _always_)
2. I can put together a PR with this change reintroduced, but with comparable fixes to the ones I made on the Trino side to mitigate the potential issue and let someone with time and realistic production workloads test them out.
3. Leave the change reverted (I have no objection with this option)